environment/src/main/java/jetbrains/exodus/env/EnvironmentImpl.java

```diff
@@ -638,26 +605,6 @@ boolean flushTransaction(@NotNull final ReadWriteTransaction txn, final boolean
         return true;
     }
 
-    MetaTree holdNewestSnapshotBy(@NotNull final TransactionBase txn) {
-        return holdNewestSnapshotBy(txn, true);
-    }
-
-    MetaTree holdNewestSnapshotBy(@NotNull final TransactionBase txn, final boolean acquireTxn) {
-        if (acquireTxn) {
-            acquireTransaction(txn);
-        }
-        final Runnable beginHook = txn.getBeginHook();
-        metaReadLock.lock();
```

Comment:
Non-synchronized execution of begin-hook introduces a race condition. Transaction-safe tasks can be illegally executed after a transaction held a meta tree and before it is registered in transaction set.